news-bundle/src/Resources/contao/dca/tl_news_archive.php

```diff
@@ -300,72 +296,6 @@ public function checkPermission()
 		}
 
 		$GLOBALS['TL_DCA']['tl_news_archive']['list']['sorting']['root'] = $root;
-		$security = System::getContainer()->get('security.helper');
-
-		// Check permissions to add archives
-		if (!$security->isGranted(ContaoNewsPermissions::USER_CAN_CREATE_ARCHIVES))
-		{
-			$GLOBALS['TL_DCA']['tl_news_archive']['config']['closed'] = true;
-			$GLOBALS['TL_DCA']['tl_news_archive']['config']['notCreatable'] = true;
-			$GLOBALS['TL_DCA']['tl_news_archive']['config']['notCopyable'] = true;
-		}
-
-		// Check permissions to delete calendars
-		if (!$security->isGranted(ContaoNewsPermissions::USER_CAN_DELETE_ARCHIVES))
-		{
-			$GLOBALS['TL_DCA']['tl_news_archive']['config']['notDeletable'] = true;
-		}
-
-		$objSession = System::getContainer()->get('session');
-
-		// Check current action
-		switch (Input::get('act'))
-		{
-			case 'select':
-				// Allow
-				break;
-
-			case 'create':
-				if (!$security->isGranted(ContaoNewsPermissions::USER_CAN_CREATE_ARCHIVES))
-				{
-					throw new AccessDeniedException('Not enough permissions to create news archives.');
-				}
-				break;
-
-			case 'edit':
-			case 'copy':
-			case 'delete':
-			case 'show':
-				if (!in_array(Input::get('id'), $root) || (Input::get('act') == 'delete' && !$security->isGranted(ContaoNewsPermissions::USER_CAN_DELETE_ARCHIVES)))
-				{
-					throw new AccessDeniedException('Not enough permissions to ' . Input::get('act') . ' news archive ID ' . Input::get('id') . '.');
-				}
-				break;
-
-			case 'editAll':
-			case 'deleteAll':
-			case 'overrideAll':
-			case 'copyAll':
-				$session = $objSession->all();
-
-				if (Input::get('act') == 'deleteAll' && !$security->isGranted(ContaoNewsPermissions::USER_CAN_DELETE_ARCHIVES))
-				{
-					$session['CURRENT']['IDS'] = array();
-				}
-				else
-				{
-					$session['CURRENT']['IDS'] = array_intersect((array) $session['CURRENT']['IDS'], $root);
-				}
-				$objSession->replace($session);
```

Comment:
Explained in a different way: before each record is edited, we ask the permission voter if it is allowed and skip if not. So even if the ID is in the session, it cannot be edited because the voter checks for it. 😎